types/lib/model.d.ts

```diff
@@ -1859,77 +1912,81 @@ export abstract class Model<T = any, T2 = any> extends Hooks {
    * profiles will be counted
    */
   public static findAndCountAll<M extends Model>(
-    this: { new(): M } & typeof Model,
-    options?: FindAndCountOptions
+    this: ModelStatic<M>,
+    options?: FindAndCountOptions<M['_attributes']>
   ): Promise<{ rows: M[]; count: number }>;
 
   /**
    * Find the maximum value of field
    */
-  public static max<M extends Model, T extends DataType | unknown>(
-    this: { new(): M } & typeof Model,
-    field: keyof M,
-    options?: AggregateOptions<T>
+  public static max<T extends DataType | unknown, M extends Model>(
+    this: ModelStatic<M>,
+    field: keyof M['_attributes'],
+    options?: AggregateOptions<T, M['_attributes']>
   ): Promise<T>;
 
   /**
    * Find the minimum value of field
    */
-  public static min<M extends Model, T extends DataType | unknown>(
-    this: { new(): M } & typeof Model,
-    field: keyof M,
-    options?: AggregateOptions<T>
+  public static min<T extends DataType | unknown, M extends Model>(
+    this: ModelStatic<M>,
+    field: keyof M['_attributes'],
+    options?: AggregateOptions<T, M['_attributes']>
   ): Promise<T>;
 
   /**
    * Find the sum of field
    */
-  public static sum<M extends Model, T extends DataType | unknown>(
-    this: { new(): M } & typeof Model,
-    field: keyof M,
-    options?: AggregateOptions<T>
+  public static sum<T extends DataType | unknown, M extends Model>(
+    this: ModelStatic<M>,
+    field: keyof M['_attributes'],
+    options?: AggregateOptions<T, M['_attributes']>
   ): Promise<number>;
 
   /**
    * Builds a new model instance. Values is an object of key value pairs, must be defined but can be empty.
    */
   public static build<M extends Model>(
-    this: { new(): M } & typeof Model,
-    record?: object,
+    this: ModelStatic<M>,
+    record?: M['_creationAttributes'],
     options?: BuildOptions
   ): M;
 
   /**
    * Undocumented bulkBuild
    */
   public static bulkBuild<M extends Model>(
-    this: { new(): M } & typeof Model,
-    records: object[],
+    this: ModelStatic<M>,
+    records: (M['_creationAttributes'])[],
     options?: BuildOptions
   ): M[];
 
   /**
    * Builds a new model instance and calls save on it.
    */
   public static create<M extends Model>(
-    this: { new(): M } & typeof Model,
-    values?: object,
-    options?: CreateOptions
+    this: ModelStatic<M>,
+    values?: M['_creationAttributes'],
```

Comment:
It is possible to create record with `id` i.e. primary key. Distinction between `CreationAttributes` and `Attributes` from user point of view make sense, but internally all attributes are accepted.

We can also think about accepting `TAttributes` generic which contains all attributes, then create a dynamic type for "CreationAttributes" by making primary key attribute optional in that type

Reply:
This is an interesting idea! The problem is - how do we figure out what's the primary key? We'd probably need users to provide it in a new generic (maybe `TPrimaryKey`) but then that's just one more thing to provide. And then, there are still fields with `defaultValue` which users can also omit, and other cases we'd need to handle.

I think `TCreationAttributes` is a pretty flexible solution that solves for both primary keys and other columns with default values that can be omitted. It's also been used for the previous sequelize v4 typings, so would be familiar to those users.

Reply:
Right, there is no good way to infer primary key in current Typescript schema. Actual attributes are passed to `init` or `define`, which contains this info.

Reply:
Yeah, I think defining it in `init` is especially problematic, so I recommend just leaving it as-is.